src/os/bluestore/aio.cc

```diff
@@ -34,7 +35,7 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
   assert(aios_size >= left);
   int done = 0;
   while (left > 0) {
-    int r = io_submit(ctx, left, piocb + done);
+    int r = io_submit(ctx, std::min(left, max_iodepth), piocb + done);
```

Comment:
Probably it makes sense to reset retries and delay once we get success. Otherwise the error might still be returned if total amount to submit exceeds 16*max_iodepth.

Reply:
Not sure whether I've got you. Even when the number of iocbs in a turn is limited by `max_iodepth`, it still wouldn't be a failure (`r < 0`) because of exceeding resources (`-EAGAIN`), and thus `attempts` shouldn't be touched at all.

Reply:
Hmm, there is a possibility to get `-EAGAIN` even when `io_submit` puts at once no more than `max_iodepth` iocbs. This could happen a result of slow drainage in kernel. Yeah, might be worth resetting.

Reply:
yep, you feed max_iodepth requests and then immediately submit the next portion on the next loop iteration and it fails due to still busy kernel...